lib/pbench/test/unit/agent/test_tool_meister.py

```diff
@@ -0,0 +1,1049 @@
+    args: List[str], tool_dir: MockedPath, ctx_name: str = None
+) -> MockCreatedProcess:
+    """Mock the behavior of Tool._create_process_with_logger() by just
+    recording what arguments we given in a tracking object.
+
+    Returns a tracking object as the "process".
+    """
+    return MockCreatedProcess(args, tool_dir, ctx_name)
+
+
+class MockProcessForTerminate:
+    """A mock for process objects which just need to know if its
+    ".terminate()" method was called.
+    """
+
+    def __init__(self, raise_exc: bool = False):
+        self.terminate_called = False
+        self.raise_exc = raise_exc
+
+    def terminate(self):
+        self.terminate_called = True
+        if self.raise_exc:
+            raise Exception("Terminate")
+
+
+class TestTool:
+    @staticmethod
+    def test_constructor_and_interfaces():
+        mocked_install_dir = MockedPath()
+        mocked_logger = NullObject()
+        tool = Tool("test-tool", "opt1;opt2", mocked_install_dir, mocked_logger)
+        assert tool._tool_type is None
+
+        with pytest.raises(NotImplementedError):
+            tool.install()
+        with pytest.raises(NotImplementedError):
+            tool.start(MockedPath())
+        with pytest.raises(NotImplementedError):
+            tool.stop()
+        with pytest.raises(NotImplementedError):
+            tool.wait()
+
+        pi_dir = MockedPath(exists=False)
+        with pytest.raises(ToolException) as excinfo:
+            Tool(name="who", tool_opts="cares", pbench_install_dir=pi_dir, logger=None)
+        assert (
+            str(excinfo.value)
+            == f"pbench installation directory does not exist: {pi_dir}"
+        )
+
+    @staticmethod
+    def test_create_process_with_logger(monkeypatch):
+        # First we need a mock of the subprocess.Popen() constructor, which
+        # will just record what the constructor was called with.
+        class MockPopen:
+            def __init__(self, args, cwd=None, stdin=None, stdout=None, stderr=None):
+                self.args = args
+                self.cwd = cwd
+                self.stdin = stdin
+                self.stdout = stdout
+                self.stderr = stderr
+
+        monkeypatch.setattr("subprocess.Popen", MockPopen)
+
+        # Next we need a mock of the threading.Thread() constructor, which
+        # will also just record what it was called with, and also offer the
+        # "daemon" field, and a "start()" method to be called.
+        the_thread = None
+
+        class MockThread:
+            def __init__(self, target: Any = None, args: Any = None):
+                self.target = target
+                self.args = args
+                self.daemon = False
+                self.start_called = False
+                nonlocal the_thread
+                the_thread = self
+
+            def start(self):
+                self.start_called = True
+
+        monkeypatch.setattr("threading.Thread", MockThread)
```

Comment:
Interesting that you mocked the `Thread` class symbol rather than mocking `__init__` to return an instance of your `MockThread`, which would have let you avoid the `nonlocal`. Or did that not work for some reason? I've done similar things with other classes...

Reply:
That sounds intriguing but I am not sure I follow how that helps.  What follows is my attempt to reason out what is possible with the given constraints, it is not trying to explain to you or anybody else what is happening.

The `_create_process_with_logger()` method is used internally by the various `Tool` class hierarchy methods to create a sub-process with a thread reading from the sub-process's `stdout` PIPE.

So we have to verify that, 1. `subprocess.Popen()` is called with the correct arguments; 2. `threading.Thread()` is called with the correct arguments; 3. a "child" logger is properly created; 4. the `daemon` filed of the `Thread` object is set to `True`; 5. the `.start()` method of the `Thread` object is invoked.

For the `process` object we construct a mock so that we can check the returned argument from `_create_process_with_logger()` is setup as expected.

But the `Thread` object is not returned in any capacity.

So something has to record the thread object and after the `_create_process_with_logger()` exists, verify that those behaviors took place.

So we _could_ mock out `threading.Thread.__init__` to return a pre-created `MockThread` object and just reference that pre-created object after `_create_process_with_logger()` returns.  However, that appears to be just a difference to remove the use of `nonlocal`.

Reply:
Nevermind; it's a subtle distinction, mostly focused on losing the `nonlocal` which is kinda ugly. I probably shouldn't even have said anything this round; the PR has gone on long enough and we should finish it. 🍤